### PR TITLE
chore: Refactor dashboard security access

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

Whilst spelunking through the codebase recently in order to fix some dashboard RBAC issues I noticed the `raise_for_dashboard_access` method (drafted prior to https://github.com/apache/superset/pull/13235) which coincides with the more generic `raise_for_access` method.

This PR extends the previous intent to migrate the logic into the `raise_for_access` method which ensures all entity based access checks use the same call stack. Furthermore this approach allows for more context aware checks, i.e., whether a user has access to a dataset/datasource in the context of viewing a dashboard.

### TESTING INSTRUCTIONS

CI.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/errors.py [hunk 1]
```diff
@@ -26,7 +26,6 @@ class SupersetErrorType(str, Enum):
     Types of errors that can exist within Superset.
 
     Keep in sync with superset-frontend/src/components/ErrorMessage/types.ts
-    and docs/src/pages/docs/Miscellaneous/issue_codes.mdx
```
Comment (john-bodley) [inline superset/errors.py:29]: These docs no longer exist.

File: superset/security/manager.py [hunk 2]
```diff
@@ -1852,6 +1867,45 @@ def raise_for_access(
                     self.get_datasource_access_error_object(datasource)
                 )
 
+        if dashboard:
+            if self.is_guest_user():
```
Comment (john-bodley) [inline superset/security/manager.py:1871]: Identical logic which was previously in `raise_for_dashboard_access`.

**Comment by john-bodley** [comment]
@eschutho and @Vitor-Avila as mentioned in https://github.com/apache/superset/pull/24892 this refactor should help with your efforts in readdressing the problem outlined in https://github.com/apache/superset/pull/24808 as now the `raise_for_access` method is dashboard (context) aware, i.e., the caller could be mutated to be of the form,

```python
security_manager.raise_for_access(dashboard=dashboard, datasource=datasource)
```

as opposed to

```python
security_manager.raise_for_access(datasource=datasource)
```

or similar.

**Comment by Vitor-Avila** [comment]
thanks @john-bodley, that makes perfect sense! I see the refactoring is now merged (along with the revert) so I can update my fork and take a look at this.

**Review by michael-s-molina — APPROVED** [review]
LGTM

**Pull request merged**